packages/contracts-bedrock/src/universal/OptimismMintableERC20FactoryInterop.sol

```
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import { OptimismMintableERC20Factory } from "./OptimismMintableERC20Factory.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { Unauthorized } from "src/libraries/errors/CommonErrors.sol";
+
+/// @custom:proxied true
+/// @custom:predeployed 0x4200000000000000000000000000000000000012
+/// @title OptimismMintableERC20FactoryInterop
+/// @notice OptimismMintableERC20FactoryInterop is an extension of the OptimismMintableERC20Factory that handles the
+/// `deployments` mapping inclusion of the tokens deployed before holocene. It uses a hash onion structure to verify and
+/// store the deployments, and can be consired complete when the hash onion is fully peeled.
+contract OptimismMintableERC20FactoryInterop is OptimismMintableERC20Factory {
+    /// @notice Thwown when the hashOnion is already set.
+    error HashOnionAlreadySet();
+
+    /// @notice Thrown when the hashOnion is not set and `verifyAndStore` is called.
+    error HashOnionNotSet();
+
+    /// @notice Thrown when the hashOnion is already peeled - meaning that all the deployments are already stored.
+    error OnionAlreadyPeeled();
+
+    /// @notice Thrown when the tokens length to verify and store mismatch.
+    error TokensLengthMismatch();
+
+    /// @notice Thrown when the computed hash onion with the given tokens and starting inner layer doesn't match the
+    /// hashOnion.
+    error InvalidHashOnion();
```

This should be named differently. Written like this, it seems like it should throw when admin trys to update the hash onion and its an invalid type or length. I would rename to `InvalidProof` or something more explicit